lua/sim/Matchstate/AbstractVictoryCondition.lua

```diff
@@ -0,0 +1,230 @@
+        for k, aiBrain in ArmyBrains do
+            if not (ArmyIsOutOfGame(aiBrain:GetArmyIndex() or aiBrain:IsDefeated())) then
+                TableInsert(participatingArmyBrains, aiBrain)
+            end
+        end
+
+        return participatingArmyBrains
+    end,
+
+    --- Utility function to help determine if all the given brains are allied and are requesting allied victory.
+    ---@param self AbstractVictoryCondition
+    ---@param aiBrains AIBrain[]
+    ---@return boolean
+    RemainingBrainsAreAllied = function(self, aiBrains)
+        local victorious = true
+        for a = 1, TableGetn(aiBrains) do
+            local aBrain = aiBrains[a]
+            local aIndex = aBrain:GetArmyIndex()
+
+            for b = 1, TableGetn(aiBrains) do
+                local bBrain = aiBrains[b]
+                local bIndex = bBrain:GetArmyIndex()
+
+                if aIndex ~= bIndex then
+                    victorious = victorious and IsAlly(aIndex, bIndex) and aBrain.RequestingAlliedVictory and bBrain.RequestingAlliedVictory
+                end
+            end
+        end
+
+        return victorious
+    end,
+
+    --- Utility function to help determine if all the given brains want to draw.
+    ---@param self AbstractVictoryCondition
+    ---@param aiBrains AIBrain[]
+    ---@return boolean
+    RemainingBrainsForfeit = function(self, aiBrains)
+        for k = 1, TableGetn(aiBrains) do
+            local brain = aiBrains[k]
+            if not brain.OfferingDraw then
+                return false
+            end
+        end
+
+        return true
+    end,
+
+    ---@param self AbstractVictoryCondition
+    Setup = function(self)
+        self.Trash:Add(ForkThread(self.ProcessGameStateThread, self))
+    end,
+
+    ---@param self AbstractVictoryCondition
+    ProcessGameStateThread = function(self)
+        while not IsGameOver() do
+            self:ProcessGameState(ArmyBrains)
+            WaitTicks(4)
+        end
+    end,
+
+    ---@param self AbstractVictoryCondition
+    ---@param aiBrains AIBrain[]
+    ProcessGameState = function(self, aiBrains)
+        error("Missing implementation of ProcessGameState")
+    end,
+
+    --- Ends the game by starting a thread that ends the game 3 seconds later.
+    ---@param self AbstractVictoryCondition
+    EndGame = function(self)
+        self.Trash:Add(ForkThread(self.EndGameThread, self))
+    end,
+
+    --- Ends the game.
+    ---@param self AbstractVictoryCondition
+    EndGameThread = function(self)
+        WaitTicks(30)
+
+        for _, v in GameOverListeners do
+            pcall(v)
+        end
+
+        Sync.GameEnded = true
+        WaitTicks(1)
+        EndGame()
+    end,
+
+    --- Utility function for the logic to remove an army from the game by turning it into an observer.
+    ---@param self AbstractVictoryCondition
+    ---@param aiBrain AIBrain
+    ToObserver = function(self, aiBrain)
+        local brainIndex = aiBrain:GetArmyIndex()
```

Comment:
```suggestion
        local brainIndex = aiBrain.Army
```
(and for all other `aiBrain:GetArmyIndex()` calls too)

Reply:
Applied in https://github.com/FAForever/fa/pull/6766/commits/1387401439bdd6bc0afc6e90088d3291e0bd1561